Spark 3.3: support write to WAP branch

jackye1995 opened the pull request:
Fixes #6774

based on #6965 , separated PR for writing to WAP branch. Will rebase once that is merged.

@rdblue @aokolnychyi @amogh-jahagirdar @namrathamyske

jackye1995 commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkSessionProperties.java:
```diff
@@ -0,0 +1,28 @@
+ */
+package org.apache.iceberg.spark;
+
+public class SparkSessionProperties {
```
Seems like we have been hard-coding `spark.wap.id`, maybe it's better to have a class to hold these commonly used session properties?

aokolnychyi replied:
Can we use existing `SparkSQLProperties`?

rdblue replied:
+1 for `SparkSQLProperties`.

jackye1995 commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestPartitionedWritesToWAPBranch.java:
```diff
@@ -0,0 +1,118 @@
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestPartitionedWritesToWAPBranch extends PartitionedWritesTestBase {
```
I did not add `TestUnpartitionedWritesToWAPBranch`, seems unnecessary to add more tests because it does not test additional case in the code path.

jackye1995 commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java:
```diff
@@ -937,6 +938,28 @@ public static org.apache.spark.sql.catalyst.TableIdentifier toV1TableIdentifier(
     return org.apache.spark.sql.catalyst.TableIdentifier.apply(table, database);
   }
 
+  public static String determineWriteBranch(
```
This method can technically be embedded in `SparkWriteConf.branch()`, and we can also remove `SparkWriteConf.wapBranch()` method. The only issue is that it will compute `wapId` and `wapEnabled` twice, once in the method, once in `SparkWrite`/`SparkPositionDeltaWrite`. Please let me know if there is any preference for code cleanness vs computation.

rdblue replied:
I think it makes more sense to embed this in `SparkWriteConf.branch()`. `wapId` and `wapEnabled` are constant so it should be fine.

amogh-jahagirdar replied:
Yeah just took a look in the code, there's no real computation for `wapId` and `wapEnabled` they're just constants, so I'm in favor of having it in `SparkWriteConf` as well.

aokolnychyi commented:
I'll take a look on Monday.

rdblue commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java:
```diff
@@ -1066,6 +1068,50 @@ public void testDeleteWithMultipleSpecs() {
 
+  @Test
+  public void testDeleteToWAPBranch() throws NoSuchTableException {
+    Assume.assumeTrue("WAP branch only works for table identifier without branch", branch == null);
```
Do we fail when writing to a branch identifier if the `WAP_BRANCH` is set? If so, we could test that behavior here instead of using `Assume`.

rdblue replied:
I see that this is tested in the next case, so we're good.

rdblue commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java:
```diff
@@ -1066,6 +1068,50 @@ public void testDeleteWithMultipleSpecs() {
+    append(new Employee(0, "hr"), new Employee(1, "hr"), new Employee(2, "hr"));
+
+    try {
+      spark.conf().set(SparkSessionProperties.WAP_BRANCH, "wap");
```
Don't we have a helper to run a lambda with this try/finally to set a property?

jackye1995 replied:
Good point, I copied this from other tests, seems like we can also update those tests with the helper.

rdblue commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java:
```diff
@@ -1066,6 +1068,50 @@ public void testDeleteWithMultipleSpecs() {
+      spark.conf().set(SparkSessionProperties.WAP_BRANCH, "wap");
+      sql("DELETE FROM %s t WHERE id=0", tableName);
+      Assert.assertEquals(
+          "Should have expected num of rows", 2L, spark.table(tableName + ".branch_wap").count());
```
Is this branch read necessary or is it the default in WAP? I think it needs to be the default so that dynamic pruning works. That said, it does make sense to allow reading specific branches, and to read a specific branch here to validate behavior.

Should we also validate that `tableName + ".branch_main"` has not been modified?

jackye1995 replied:
added

rdblue commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java:
```diff
@@ -1066,6 +1068,50 @@ public void testDeleteWithMultipleSpecs() {
+          .isInstanceOf(ValidationException.class)
+          .hasMessage(
+              String.format(
+                  "Cannot set both branch and WAP branch, but got branch [%s] and WAP branch [wap]",
```
"Cannot write to" isntead of "Cannot set"?

rdblue commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java:
```diff
@@ -2448,6 +2450,65 @@ public void testMergeNonExistingBranch() {
+      assertEquals(
+          "Should correctly merge to WAP branch",
+          expectedRows,
+          sql("SELECT * FROM %s.branch_wap ORDER BY id", tableName));
```
Same general comments as above. It would be nice to validate both main and the "wap" branch.

rdblue commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestPartitionedWritesToWAPBranch.java:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  // commit target in WAP is just the table name
+  // should use table + branch name instead for read
```
I don't agree with this. I think it has to read the branch if it exists.

For example, a DELETE command is going to modify the state of a branch if it exists, and it could exist because branch WAP supports multiple writes. That means the reads for both the delete itself and any dynamic pruning must read the branch if it exists and main if it doesn't.

amogh-jahagirdar replied:
Yeah we need to use the `branch` during planning of the write to handle the dynamic pruning case, that's what came up in https://github.com/apache/iceberg/pull/6651/files. If we pass the branch through the SparkTable that should take care of this? Whether it's WAP branch or not doesn't matter, just needs to read the branch

amogh-jahagirdar commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestPartitionedWritesToWAPBranch.java:
```diff
@@ -0,0 +1,118 @@
+    assertEquals(
+        "View should have expected rows", ImmutableList.of(row(1L, "a")), sql("SELECT * FROM tmp"));
+
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", commitTarget());
```
very minor nit: If we could make the second insert, insert another value for the second column it makes it a bit more obvious this is another record just from reading the expectation. but very minor, just something to make the test more apparent imo.

jackye1995 commented:
@rdblue @amogh-jahagirdar thanks for the suggestions, I have addressed all nit comments, and also fixed the scan issue and updated the tests to verify behavior of multiple writes. Could you take another look?

jackye1995 commented:
> Should we also validate that tableName + ".branch_main" has not been modified?

Overlooked this one, will add some lines to verify in a few hours.

amogh-jahagirdar commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java:
```diff
@@ -1066,6 +1068,73 @@ public void testDeleteWithMultipleSpecs() {
+              "Should have expected num of rows when reading WAP branch with multiple writes",
+              1L,
+              spark.table(tableName + ".branch_wap").count());
+          Assert.assertEquals(
+              "Should not modify main branch with multiple writes",
+              3L,
+              spark.table(tableName + ".branch_main").count());
```
Nit: I think for the purpose of this test we could just do `spark.table(tableName).count()` instead of main, don't think we need to validate the identifier read here.

amogh-jahagirdar replied:
Ah nvm, I misinterpreted, we actually do need this!

amogh-jahagirdar reviewed (approved):
I took another pass @jackye1995 just a nit, this looks great to me overall!

rdblue commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:
```diff
@@ -333,6 +334,28 @@ public boolean caseSensitive() {
+            branch == null,
+            "Cannot write to both branch and WAP branch, but got branch [%s] and WAP branch [%s]",
+            branch,
+            wapBranch);
```
I don't think this behavior is a blocker because it is strict, but I would expect to be able to write to another branch with the WAP branch set. I'm curious what other people think the long-term behavior should be.

I think this behavior does help ensure that there are no side-effects, which is good if you want people to trust the pattern. But that's undermined by enabling/disabling WAP on a per-table basis.

jackye1995 replied:
Thank you! I added issue https://github.com/apache/iceberg/issues/7103 and we can discuss there with related people.

aokolnychyi replied:
It seems like a reasonable starting point to me.

rdblue commented:
Thanks @jackye1995! Looks great.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkReadConf.java:
```diff
@@ -98,7 +98,22 @@ public String branch() {
+      return inputBranch;
+    }
+
+    boolean wapEnabled =
```
nit: I'd prefer a separate method called `wapEnabled()` like we have in SparkWriteConf. Then we could use the constant for the default value and it would simplify this method.

```
public boolean wapEnabled() {
  return confParser
      .booleanConf()
      .tableProperty(TableProperties.WRITE_AUDIT_PUBLISH_ENABLED)
      .defaultValue(TableProperties.WRITE_AUDIT_PUBLISH_ENABLED_DEFAULT)
      .parse();
}
```

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:
```diff
@@ -333,6 +334,28 @@ public boolean caseSensitive() {
+    if (wapEnabled()) {
+      String wapId = wapId();
+      String wapBranch =
+          confParser.stringConf().sessionConf(SparkSQLProperties.WAP_BRANCH).parseOptional();
```
nit: What about a separate method like we have for `wapId()`?

aokolnychyi reviewed (commented):
Late +1 from me too.
